### PR TITLE
Raw and Coerced Variables refactor

### PR DESCRIPTION
**Pull request by dondonz**
# Why refactor? Identifying double coercion
Variables were previously represented as a `Map<String, Object>`, which made it difficult to determine if variables were raw or had been already coerced.

## Example: double coercion in `FieldValidationInstrumentation`

Let's start from this line in `Execution#executeOperation`. At this point, variables have already been coerced. The `ExecutionContext` can only contain coerced variables.
```
InstrumentationContext<ExecutionResult> executeOperationCtx = instrumentation.beginExecuteOperation(instrumentationParams);
```

Then inside `FieldValidationInstrumentation#beginExecuteOperation` we have
```
List<GraphQLError> errors = FieldValidationSupport.validateFieldsAndArguments(fieldValidation, parameters.getExecutionContext());
```
Within `validateFieldsAndArguments`, we create a `QueryTraverser`, which coerces variables a **second** time.

## Example: double coercion in WIP PR to shift max query depth after validation
https://github.com/graphql-java/graphql-java/pull/2773/files#diff-2696e88b9e4d095eea4d0e47b1c98ea872bfce3642c5aaee7317b1c3c2c478a5R149

This is a WIP PR so I don't want to be harsh, nevertheless it's a good illustration of how confusing raw and coerced variables are. The `ExecutionContext` can only contain coerced variables. The linked `QueryTraverser` will coerce variables a **second** time. 

# This PR adds `RawVariables` and `CoercedVariables`
 This PR adds `RawVariables` and `CoercedVariables` to make the distinction clearer and avoid accidental double coercion.

This PR addresses #2764, and sets the foundation for #2773.

This is a large PR! To make it easier to see the changes, please read commit by commit (or alternatively, I can split this work into multiple PRs).

## What did change
I refactored variables to `RawVariables` or `CoercedVariables` in crucial classes related to Execution and classes required for the upcoming traverser changes in #2773. 

## What didn't change
There are other classes containing a map of variables such as `FieldCollectionParameters`. These classes have been left unchanged primarily to keep the scope of this PR manageable. In these unchanged classes, there was far less ambiguity about coercion. I can extend `RawVariables` and `CoercedVariables` to all other classes in another PR.

## What is in the next PR
After this PR is merged in I can help finish off #2773 and fix both double coercion examples I outlined above.

### CODE REVIEW

File: src/main/java/graphql/ExecutionInput.java [hunk 1]
```diff
@@ -39,7 +40,7 @@ private ExecutionInput(Builder builder) {
         this.context = builder.context;
         this.graphQLContext = assertNotNull(builder.graphQLContext);
         this.root = builder.root;
-        this.variables = builder.variables;
+        this.rawVariables = builder.rawVariables;
```
Comment (dondonz) [inline src/main/java/graphql/ExecutionInput.java:43]: Variables in `ExecutionInput` are always raw

File: src/main/java/graphql/analysis/QueryTraverser.java [hunk 2]
```diff
@@ -42,26 +44,26 @@ public class QueryTraverser {
...
-        assertNotNull(document, () -> "document  can't be null");
+        assertNotNull(document, () -> "document can't be null");
         NodeUtil.GetOperationResult getOperationResult = NodeUtil.getOperation(document, operation);
         List<VariableDefinition> variableDefinitions = getOperationResult.operationDefinition.getVariableDefinitions();
         this.schema = assertNotNull(schema, () -> "schema can't be null");
         this.fragmentsByName = getOperationResult.fragmentsByName;
         this.roots = singletonList(getOperationResult.operationDefinition);
         this.rootParentType = getRootTypeFromOperation(getOperationResult.operationDefinition);
-        this.variables = coerceVariables(assertNotNull(variables, () -> "variables can't be null"), variableDefinitions);
+        this.coercedVariables = coerceVariables(assertNotNull(variables, () -> "variables can't be null"), variableDefinitions);
```
Comment (dondonz) [inline src/main/java/graphql/analysis/QueryTraverser.java:62]: For now, keep functionality similar to before

In the next PR (#2773), `QueryTraverser` will change to accept already coerced variables

File: src/main/java/graphql/execution/CoercedVariables.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import graphql.collect.ImmutableMapWithNullValues;
+
+import java.util.Map;
+
+/**
+ * Holds coerced variables
+ */
+@Internal
+public class CoercedVariables {
+    private final ImmutableMapWithNullValues<String, Object> coercedVariables;
```
Comment (dondonz) [inline src/main/java/graphql/execution/CoercedVariables.java:13]: Using this special map as variable values can be null

File: src/main/java/graphql/execution/ExecutionContext.java [hunk 4]
```diff
@@ -43,7 +42,7 @@ public class ExecutionContext {
     private final ImmutableMap<String, FragmentDefinition> fragmentsByName;
     private final OperationDefinition operationDefinition;
     private final Document document;
-    private final ImmutableMapWithNullValues<String, Object> variables;
+    private final CoercedVariables coercedVariables;
```
Comment (dondonz) [inline src/main/java/graphql/execution/ExecutionContext.java:45]: `ExecutionContext` can only contain coerced variables. This change makes it clearer

File: src/main/java/graphql/execution/RawVariables.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import graphql.collect.ImmutableMapWithNullValues;
+
+import java.util.Map;
+
+/**
+ * Holds raw variables, which not have been coerced yet
+ */
+@Internal
+public class RawVariables {
+    private final ImmutableMapWithNullValues<String, Object> rawVariables;
```
Comment (dondonz) [inline src/main/java/graphql/execution/RawVariables.java:13]: Using this special map as values can be null

File: src/test/groovy/graphql/execution/ConditionalNodesTest.groovy [hunk 6]
```diff
@@ -1,25 +1,21 @@
...
         given:
         def variables = new LinkedHashMap<String, Object>()
         ConditionalNodes conditionalNodes = new ConditionalNodes()
-        conditionalNodes.valuesResolver = Mock(ValuesResolver)
 
         def argument = Argument.newArgument("if", new BooleanValue(true)).build()
         def directives = [Directive.newDirective().name("skip").arguments([argument]).build()]
 
-        conditionalNodes.valuesResolver.getArgumentValues(Directives.SkipDirective.getArguments(), [argument], variables) >> ["if": true]
-
```
Comment (dondonz) [inline src/test/groovy/graphql/execution/ConditionalNodesTest.groovy:22]: This mock wasn't necessary

File: src/test/groovy/graphql/execution/ValuesResolverTest.groovy [hunk 7]
```diff
@@ -39,24 +39,22 @@ class ValuesResolverTest extends Specification {
...
     def "getVariableValues: simple variable input #inputValue"() {
         given:
         def schema = TestUtil.schemaWithInputType(inputType)
         VariableDefinition variableDefinition = new VariableDefinition("variable", variableType, null)
         when:
-        def resolvedValues = resolver.coerceVariableValues(schema, [variableDefinition], [variable: inputValue])
+        def resolvedValues = resolver.coerceVariableValues(schema, [variableDefinition], new RawVariables([variable: inputValue]))
         then:
-        resolvedValues['variable'] == outputValue
+        resolvedValues.get('variable') == outputValue
```
Comment (dondonz) [inline src/test/groovy/graphql/execution/ValuesResolverTest.groovy:50]: Most API changes in this test file

File: src/main/java/graphql/execution/ExecutionContextBuilder.java [hunk 8]
```diff
@@ -39,7 +39,7 @@ public class ExecutionContextBuilder {
     Object root;
     Document document;
     OperationDefinition operationDefinition;
-    ImmutableMapWithNullValues<String, Object> variables = ImmutableMapWithNullValues.emptyMap();
+    CoercedVariables coercedVariables = new CoercedVariables(Collections.emptyMap());
```
Comment (bbakerman) [inline src/main/java/graphql/execution/ExecutionContextBuilder.java:42]: how about `CoercedVariables.emptyVariables()` - reads nicer - not essential

File: src/main/java/graphql/execution/CoercedVariables.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+    public CoercedVariables(Map<String, Object> coercedVariables) {
+        this.coercedVariables = ImmutableMapWithNullValues.copyOf(coercedVariables);
+    }
+
+    public Map<String, Object> getMap() {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/CoercedVariables.java:19]: map name it `toMap` that is you are converting it from a X -> Y - from a CoercedVariables to a map representtion

File: src/test/groovy/graphql/ExecutionInputTest.groovy [hunk 10]
```diff
@@ -111,6 +135,31 @@ class ExecutionInputTest extends Specification {
...
+        then:
+        executionInput.graphQLContext == graphQLContext
+        executionInput.root == root
+        executionInput.rawVariables == rawVariables
+        executionInput.dataLoaderRegistry == registry
+        executionInput.cacheControl == cacheControl
+        executionInput.locale == Locale.GERMAN
+        executionInput.extensions == [some: "map"]
+        executionInput.query == "new query"
+    }
```
Comment (bbakerman) [inline src/test/groovy/graphql/ExecutionInputTest.groovy:161]: What happens if we set both

```
.rawVariables(x).variables(y)
```

Should we have a test for this.  Do they go into the same place?

**Review by bbakerman — APPROVED** [review]
I generally appove strongly of this

See my specific comments on naming.

File: src/test/groovy/graphql/ExecutionInputTest.groovy (hunk 10 above)
Comment (dondonz) [reply src/test/groovy/graphql/ExecutionInputTest.groovy:161]: I'll add a test.

Yes, in `ExecutionInput` both `rawVariables` and `variables` set the same field. If both get set, the last one would be used (in your example `y`).

Is it worthwhile preventing both `.rawVariables(x).variables(y)` to be used? I don't mind adding another check if it'll be clearer

**Review by bbakerman — CHANGES_REQUESTED** [review]
As we discussed in person - lets remove the RawVariables fron ExecutionInput signature but use it on the inside

**Review by bbakerman — APPROVED** [review]
Merge !   Merge !   Merge!!

**Pull request merged**